frontend/api/auth/index.ts

```diff
@@ -0,0 +1,217 @@
+import passport from 'passport';
+import { sign } from 'jsonwebtoken';
+import express, { Express, Request, Response, NextFunction } from 'express';
+import { Strategy as FacebookStrategy } from 'passport-facebook';
+import { Strategy as GoogleStrategy } from 'passport-google-oauth20';
+import { Strategy as TwitterStrategy } from 'passport-twitter';
+import type { Profile } from 'passport';
+import { prisma } from '../../src/server/prisma';
+import { assignAnswerToUser } from '../answers';
+import ms from 'ms';
+import { respond404 } from '../../src/server/errors';
+
+const app: Express = express();
+
+if (process.env.VERCEL_ENV === 'preview') {
+  process.env[
+    'BASE_URL'
+  ] = `https://volebni-kalkulacka-2022-git-${process.env.VERCEL_GIT_COMMIT_REF}-ceskodigital.vercel.app`;
+}
+
+const BASE_URL = process.env['BASE_URL'] || 'http://localhost:3000';
+
+const providers = {
+  facebook: {
+    strategy: () => {
+      return new FacebookStrategy(
+        {
+          clientID: process.env['FACEBOOK_CLIENT_ID'] as string,
+          clientSecret: process.env['FACEBOOK_CLIENT_SECRET'] as string,
+          profileFields: ['id', 'displayName', 'email', 'gender'],
+          callbackURL: `${BASE_URL}/api/auth/facebook/callback`,
+        },
+        getStrategyCallback('facebook')
+      );
+    },
+    enabled: !!(
+      process.env['FACEBOOK_CLIENT_ID'] && process.env['FACEBOOK_CLIENT_SECRET']
+    ),
+  },
+  google: {
+    strategy: () => {
+      return new GoogleStrategy(
+        {
+          clientID: process.env['GOOGLE_CLIENT_ID'] as string,
+          clientSecret: process.env['GOOGLE_CLIENT_SECRET'] as string,
+          callbackURL: `${BASE_URL}/api/auth/google/callback`,
+        },
+        getStrategyCallback('google')
+      );
+    },
+    enabled: !!(
+      process.env['GOOGLE_CLIENT_ID'] && process.env['GOOGLE_CLIENT_SECRET']
+    ),
+  },
+  twitter: {
+    strategy: () => {
+      return new TwitterStrategy(
+        {
+          consumerKey: process.env['TWITTER_CONSUMER_KEY'] as string,
+          consumerSecret: process.env['TWITTER_CONSUMER_SECRET'] as string,
+          callbackURL: `${BASE_URL}/api/auth/twitter/callback`,
+        },
+        getStrategyCallback('twitter')
+      );
+    },
+    enabled: !!(
+      process.env['TWITTER_CONSUMER_KEY'] &&
+      process.env['TWITTER_CONSUMER_SECRET']
+    ),
+  },
+};
+
+const getStrategyCallback = (strategy: string) => {
+  return async (
+    accessToken: string,
+    refreshToken: string,
+    profile: Profile,
+    cb: (error: any, user?: any, info?: any) => void
+  ) => {
+    if (!profile.emails || !profile.emails.length || !profile.emails[0].value) {
+      return cb({ error: 'No email on facebook profile' }, false, {
+        message: 'No email provided',
```

Comment:
Přidáš tam teda ten telefon? Pokud je to tak, že u Facebooku musí být buď email nebo telefon.

Reply:
facebook telefon nevraci, bylo byl by to profil s prazdnym emailem a moznosti se do nej prihlasit pouze pres fb.

Reply:
Tohle jsem tam upravil, ale prosim jeste o diskuzi. @krystof-k @martinwenisch